test/induction.lean

```diff
@@ -0,0 +1,1338 @@
+  a = b :=
+begin
+  cases' a,
+  cases' b,
+  guard_hyp num : ℤ,
+  guard_hyp denom : ℤ,
+  guard_hyp num_1 : ℤ,
+  guard_hyp denom_1 : ℤ,
+  rw fraction.mk.inj_eq,
+  exact and.intro hnum hdenom
+end
+
+-- A "with" clause can be used to give the names of constructor arguments (as
+-- for `cases`, `induction` etc).
+example (x : ℕ × ℕ) (y : Vec ℕ 2) (z : List ℕ) : unit :=
+begin
+  cases' x with i j k l,
+  guard_hyp i : ℕ,
+  guard_hyp j : ℕ,
+  clear i j,
+
+  cases' y with i j k l,
+  -- Note that i is 'skipped' because it is used to name the (n : ℕ)
+  -- argument of `cons`, but that argument is cleared by index unification. I
+  -- find this a little strange, but `cases` also behaves like this.
+  guard_hyp j : ℕ,
+  guard_hyp k : Vec ℕ 1,
+  clear j k,
+
+  cases' z with i j k l,
+  case nil { exact () },
+  case cons {
+    guard_hyp i : ℕ,
+    guard_hyp j : List ℕ,
+    exact ()
+  }
+end
+
+-- "with" also works with induction'.
+example (x : List ℕ) : unit :=
+begin
+  induction' x with i j k l,
+  case nil { exact () },
+  case cons {
+    guard_hyp i : ℕ,
+    guard_hyp j : List ℕ,
+    guard_hyp k : unit,
+    exact ()
+  }
+end
+
+-- An underscore in a "with" clause means "use the auto-generated name for this
+-- argument".
+example (x : List ℕ) : unit :=
+begin
+  induction' x with _ j _ l,
+  case nil { exact () },
+  case cons {
+    guard_hyp x : ℕ,
+    guard_hyp j : List ℕ,
+    guard_hyp ih : unit,
+    exact ()
+  }
+end
+
+-- induction' and cases' can be used to perform induction/case analysis on
+-- arbitrary expressions (not just hypotheses). A very synthetic example:
+example {α} : α ∨ ¬ α :=
+begin
+  cases' classical.em α with a nota,
+  { exact (or.inl a) },
+  { exact (or.inr nota) }
+end
+
+-- Cases'/induction' can add an equation witnessing the case split they
+-- performed. Again, a highly synthetic example:
+example {α} (xs : list α)
+  : xs.reverse.length = 0 ∨ ∃ m, xs.reverse.length = m + 1 :=
+begin
+  cases' eq : xs.length,
+  case zero {
+    left,
+    rw list.length_reverse,
+    exact eq
+  },
+  case succ : l {
+    right,
+    rw list.length_reverse,
+    use l,
+    exact eq
+  }
+end
+
+-- Index equation simplification can deal with equations that aren't in normal
+-- form.
+example {α} (x : α) (xs) : list.take 1 (x :: xs) ≠ [] :=
+begin
+  intro contra,
+  cases' contra
+end
+
+-- Index generalisation should leave early occurrences of complex index terms
+-- alone. This means that given the major premise `e : E (f y) y` where `y` is a
+-- complex term, index generalisation should give us
+--
+--     e : E (f y) i,
+--
+-- *not*
+--
+--     e : E (f i) i.
+--
+-- Otherwise we get problems with examples like this:
+inductive ℕ₂ : Type
+| zero
+| succ (n : ℕ₂) : ℕ₂
+
+namespace ℕ₂
+
+def plus : ℕ₂ → ℕ₂ → ℕ₂
+| zero y := y
+| (succ x) y := succ (plus x y)
+
+example (x : ℕ₂) (h : plus zero x = zero) : x = zero :=
+begin
+  cases' h,
+  guard_target zero = plus zero zero,
```

Comment:
Why is `zero` replaced by the index variable in the goal? Wouldn't the goal be simpler if you only generalized the `zero` in the index position (the RHS of `h`)? Is there an advantage of doing that?

I thought you maybe did it to keep the goal type-correct when the goal depends on `h`, but that case is not supported:
```lean
example (x : ℕ₂) (h : plus zero x = zero) : subtype.mk x h = subtype.mk zero rfl :=
begin
  cases' h, -- error
end
```
(note: this is a very minor regression from the `cases` tactic, which does work here)

Reply:
The reason for this is indeed to prevent type errors when the goal depends on the major premise or the index arguments. This is needed for example for `head_induction_on` in the test suite:

```lean
inductive star {α : Sort*} (r : α → α → Prop) (a : α) : α → Prop
| refl {}    : star a
| tail {b c} : star b → r b c → star c

namespace star

lemma head_induction_on {α} {b : α} {r} {P : ∀a : α, star r a b → Prop} {a} (h : star r a b)
  (refl : P b refl)
  (head : ∀{a c} (h' : r a c) (h : star r c b), P c h → P a (h.head h')) :
  P a h :=
begin
  induction' h,
  ...
end

end star
```

`induction'` wants to generalise `b` to `index : α` with `index = b`. (This is useless here because `b` is not complex, but I currently don't check for that due to laziness.) However, then the hypothesis `P` also needs to be changed to `P : ∀a : α, star r a index → Prop` to keep the goal type-correct.

With that said, your example clearly demonstrates that blindly replacing every occurrence of the generalised argument is not a great idea. I'm not sure yet what the appropriate fix for this is.

Reply:
Interesting. So in that case the goal depends on the major premise and the tactic does work. Do you know what the difference with my example that fails?

Reply:
One possibility is to only replace the complex index with the variable `index` if the major premise occurs in the goal (after reverting other hypotheses). This is a pretty blunt heuristic though, and there are probably better ones. My feeling is that you want to replace the minimum number of occurrences while remaining type-correctness (or do you know of any examples where replacing a complex index with `index` leads to a simpler goal?)

Reply:
I guess the difference between your example and mine is that in my example, `b` occurs only in positions that are relevant to `h` in some sense, whereas in your example the occurrence of `zero` in the goal really has nothing to do with `h`. So maybe your intuition is correct and we should replace only those occurrences that must be replaced to preserve well-typedness. However, I don't know how to implement this other than with a brute-force approach that iterates through each possible partial replacement.

I've added your test case and a TODO note with a link to this discussion, but will merge the PR as-is for now.